### PR TITLE
Fix `force_derivation` bug

### LINKED ISSUE #1626: Derivation with `force_derivation: false` is broken
I'm afraid #1562 broke the derivation when using `force_derivation: false` (the default option). The [recipe](https://github.com/ESMValGroup/ESMValCore/files/8862393/recipe_custom_cmor_location.yml.txt) fails with the error given below. Please note that the preprocessor does not try to derive the varible from `rsds` and `rsus` as requested by the [derivation script](https://github.com/ESMValGroup/ESMValCore/blob/main/esmvalcore/preprocessor/_derive/rsns.py), but rather tries to extract the variable `rsns` directly (which does not exist).

I think this is the corresponding change that causes the error: https://github.com/ESMValGroup/ESMValCore/pull/1562/files#diff-b675431b6cf759c61f7867680472aabc0304475fd666ed514ca0365bb87c2a47R1164. We definitely need a check for the existence of the files here. However, since this is not possible if wildcards are used (which themselves would needs files to be resolved), this becomes a chicken-egg problem.

Thus, I think we should disable the use of wildcards when `force_derivation: false` is used.

```bash
2022-06-08 14:02:53,658 UTC [1544702] ERROR   No input files found for variable {'mip': 'Amon', 'derive': True, 'force_derivation': False, 'timerange': '2000/2005', 'variable_group': 'rsns_derive_input_rsns', 'short_name': 'rsns', 'diagnostic': 'map', 'preprocessor': 'default', 'dataset': 'CESM2', 'project': 'CMIP6', 'ensemble': 'r1i1p1f1', 'grid': 'gn', 'exp': 'historical', 'recipe_dataset_index': 0, 'institute': ['NCAR'], 'activity': 'CMIP', 'alias': 'CESM2', 'original_short_name': 'rsns', 'standard_name': 'surface_net_downward_shortwave_flux', 'long_name': 'Surface Net downward Shortwave Radiation', 'units': 'W m-2', 'modeling_realm': ['atmos'], 'frequency': 'mon', 'start_year': 2000, 'end_year': 2005}
2022-06-08 14:02:53,659 UTC [1544702] ERROR   Looked for files matching ['rsns_Amon_CESM2_historical_r1i1p1f1_gn*.nc'], but did not find any existing input directory
2022-06-08 14:02:53,659 UTC [1544702] ERROR   Set 'log_level' to 'debug' to get more information
2022-06-08 14:02:53,659 UTC [1544702] ERROR   Could not create all tasks
2022-06-08 14:02:53,659 UTC [1544702] ERROR   Missing data for preprocessor map/rsns_derive_input_rsns:
- Missing data for CESM2: rsns
2022-06-08 14:02:53,660 UTC [1544702] ERROR   Not all input files required to run the recipe could be found.
2022-06-08 14:02:53,660 UTC [1544702] ERROR   If the files are available locally, please check your `rootpath` and `drs` settings in your user configuration file /home/b/b309141/config/dkrz.yml
2022-06-08 14:02:53,660 UTC [1544702] ERROR   To automatically download the required files to `download_dir: /work/bd0854/DATA/ESMValTool2/download`, set `offline:
false` in /home/b/b309141/config/dkrz.yml or run the recipe with the extra command line argument --offline=False
2022-06-08 14:02:53,660 UTC [1544702] INFO    Note that automatic download is only available for files that are hosted on the ESGF, i.e. for projects: CMIP3, CMIP5, CMIP6, CORDEX, and obs4MIPs
2022-06-08 14:02:54,628 UTC [1544702] INFO    Maximum memory used (estimate): 0.0 GB
2022-06-08 14:02:54,628 UTC [1544702] INFO    Sampled every second. It may be inaccurate if short but high spikes in memory consumption occur.
2022-06-08 14:02:54,630 UTC [1544702] ERROR   Could not create all tasks
```

### PR DESCRIPTION
**Pull request by sloosvel**
## Description

This PR adds a check that does not allow to combine `force_derivation: false` and wildcards in the `timerange` tag and should the previous `force_derivation: false` behaviour.

Closes #1626 

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [ ] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [ ] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [ ] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [ ] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [ ] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [ ] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/_recipe.py [hunk 1]
```diff
@@ -1161,9 +1161,16 @@ def append(group_prefix, var):
 
     for variable in variables:
         group_prefix = variable['variable_group'] + '_derive_input_'
-        if not variable.get('force_derivation'):
+        if not variable.get('force_derivation') and \
+           '*' in variable['timerange']:
+            raise RecipeError(
+                f"Error in {variable['short_name']}: "
+                "Setting force_derivation to False and using wildcards "
+                "in time range is not allowed."
```
[the remark this answers is not in the document]
Comment (sloosvel) [reply esmvalcore/_recipe.py]: Better? Does this solve your issue?

**Review by schlunma — COMMENTED** [review]
Thanks for these quick changes Saskia, everything works fine now!!

One question: While testing this I realized that our data finder struggles to find data when the number of digits for the start year is longer than the number of digits of the end year (e.g. `timerange: 200001/2001`). The reason for this is that the comparison used to check if a file lies within the given timerange  (`start <= time_in_file <= end`) never returns anything since the integer 200001 is never smaller than the integer 2001.

I don't think we can easily fix this in a general way, as padding these numbers automatically is quite dangerous (e.g., we also have 3-digit years, and can probably also have 5-digit years, in which cases an automatic padding would be plainly wrong).

I would suggest that we add a note to the doc that users need to be carefulhere and specify the same number of digits for the start and end period. Can I simply push this to this branch here? Thanks!!

**Comment by sloosvel** [comment]
I am not sure how this is related to this pull request, but if clarifying this is useful to you, please add the comments as needed.

**Comment by schlunma** [comment]
The problem I encountered was that I got missing data errors when using `timerange: */2000` for a dataset even though the data is there. Turns out the wildcard is resolved to `197901/2000` in this case, thus leading to the problem described above.

**Comment by sloosvel** [comment]
amazing, please push whatever is needed to the branch and approve it so it can be merged!

**Review by schlunma — APPROVED** [review]

**Pull request merged**